pkgs/tools/virtualization/mkosi/default.nix

```diff
@@ -38,34 +40,51 @@ let
     withKernelInstall = true;
   };
 
-  python3pefile = python3.withPackages (ps: with ps; [
-    pefile
-  ]);
+  python3pefile = python3.withPackages (
+    ps: with ps; [
+      pefile
+    ]
+  );
 in
 buildPythonApplication rec {
   pname = "mkosi";
-  version = "22";
+  version = "24.3-unstable-2024-08-28";
   format = "pyproject";
 
-  outputs = [ "out" "man" ];
+  outputs = [
+    "out"
+    "man"
+  ];
 
   src = fetchFromGitHub {
     owner = "systemd";
     repo = "mkosi";
-    rev = "v${version}";
-    hash = "sha256-Zom1GlyhqgpTKfjcBOUEJMlubSn+TQsk97js1/UfDHY=";
+    rev = "8c2f828701a1bdb3dc9b80d6f2ab979f0430a6b8";
+    hash = "sha256-rO/4ki2nAJQN2slmYuHKESGBBDMXC/ikGf6dMDcKFr4=";
   };
 
-  # Fix ctypes finding library
-  # https://github.com/NixOS/nixpkgs/issues/7307
-  postPatch = lib.optionalString stdenv.isLinux ''
-    substituteInPlace mkosi/user.py \
-      --replace-fail 'ctypes.util.find_library("c")' "'${stdenv.cc.libc}/lib/libc.so.6'"
-    substituteInPlace mkosi/__init__.py \
-      --replace-fail '/usr/lib/systemd/ukify' "${systemdForMkosi}/lib/systemd/ukify"
-  '' + lib.optionalString withQemu ''
-    substituteInPlace mkosi/qemu.py \
-      --replace-fail "usr/share/qemu/firmware" "${qemu}/share/qemu/firmware"
+  patches =
+    [
+      (replaceVars ./0001-Use-wrapped-binaries-instead-of-Python-interpreter.patch {
+        UKIFY = "${systemdForMkosi}/lib/systemd/ukify";
+        PYTHON_PEFILE = "${python3pefile}/bin/python3.12";
+        MKOSI_SANDBOX = "~MKOSI_SANDBOX~"; # to satisfy replaceVars, will be replaced in postPatch
+      })
+      (replaceVars ./0002-Fix-library-resolving.patch {
+        LIBC = "${stdenv.cc.libc}/lib/libc.so.6";
+        LIBSECCOMP = "${libseccomp.lib}/lib/libseccomp.so.2";
+      })
+    ]
+    ++ lib.optional withQemu (
+      replaceVars ./0003-Fix-QEMU-firmware-path.patch {
+        QEMU_FIRMWARE = "${qemu}/share/qemu/firmware";
+      }
+    );
+
+  postPatch = ''
+    # As we need the $out reference, we can't use `replaceVars` here.
```

Comment:
We can: https://noogle.dev/f/builtins/placeholder

Reply:
Doesn't seem to work here, as it replaces `out` with the store path of the patch here:

```
/nix/store/pdd3zch8pljwhr9w92yv5k1gd2w5z037-0001-Use-wrapped-binaries-instead-of-Python-interpreter.patch/bin/mkosi-sandbox
```

Or can I somehow force it to evaluate in the outer derivation and then pass it down into the patch / replaceVars?

Reply:
I think it's still easier to follow along than having *that many* `substituteInPlaces`, like @philiptaron mentioned. I don't think we need to get rid of `substituteInPlace` entirely, but the whole `postPatch` block (without formatting of an kind) being minimal seems like a good trade-off to me.